### PR TITLE
XGBoost autologging: support per-class importance plots

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

XGBoost 1.15.0-dev introduced support for importance computation on linear estimators. These estimators return importance values for each (feature, class) pair as a `num_features`-by-`num_classes` matrix. This PR introduces extends feature importance plotting support in XGBoost autologging to handle this new importance value format.

## How is this patch tested?

- Unit tests
- Manual tests on XGBoost 1.15.0-dev:

1. Linear booster training on Iris

```python
from sklearn.datasets import load_iris

iris = load_iris()

import mlflow
mlflow.xgboost.autolog()


import xgboost as xgb
dtrain = xgb.DMatrix(iris.data, label=iris.target)

bst_params = {"objective": "multi:softprob", "num_class": 3, "booster": "gblinear"}
model = xgb.train(bst_params, dtrain)
```

feature_importance_weight

2. Tree booster training on Iris

```python
from sklearn.datasets import load_iris

iris = load_iris()

import mlflow
mlflow.xgboost.autolog()


import xgboost as xgb
dtrain = xgb.DMatrix(iris.data, label=iris.target)

bst_params = {
        "objective": "multi:softprob",
        "num_class": 10,
        "eval_metric": "mlogloss",
        "booster": "gbtree",
    }
model = xgb.train(bst_params, dtrain)
```

feature_importance_weight

3. Linear booster training on MNIST

```python
from sklearn.datasets import load_digits

digits = load_digits()

import mlflow
mlflow.xgboost.autolog()


import xgboost as xgb
dtrain = xgb.DMatrix(digits.data, label=digits.target)

bst_params = {"objective": "multi:softprob", "num_class": 10, "booster": "gblinear"}
model = xgb.train(bst_params, dtrain)

```

feature_importance_weight

4. Tree booster training on MNIST

```
from sklearn.datasets import load_digits

digits = load_digits()

import mlflow
mlflow.xgboost.autolog()


import xgboost as xgb
dtrain = xgb.DMatrix(digits.data, label=digits.target)

bst_params = {
        "objective": "multi:softprob",
        "num_class": 10,
        "eval_metric": "mlogloss",
        "booster": "gbtree",
    }
model = xgb.train(bst_params, dtrain)
```

feature_importance_weight

## Release Notes

Add XGBoost autologging support for multi-class feature importance plots

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/xgboost.py [hunk 1]
```diff
@@ -434,26 +434,75 @@ def log_feature_importance_plot(features, importance, importance_type):
...
+            if importances_per_class_by_feature.ndim <= 1:
+                # In this case, the supplied `importance` values are not given per class. Rather,
+                # one importance value is given per feature. For consistency with the assumed
+                # `num_features`-by-`num_classes` matrix structure, we coerce the importance
+                # values to a `num_features`-by-1 matrix
+                indices = np.argsort(importance)
+                features = features[indices]
+                importances_per_class_by_feature = np.array(
+                    [[importance] for importance in importances_per_class_by_feature]
+                )
```
Comment (harupy) [inline mlflow/xgboost.py:452]:
```suggestion
                importances_per_class_by_feature = np.array(
                    [[importance] for importance in importances_per_class_by_feature[indices]]
                )
```

Can we sort `importance` as well?

File: mlflow/xgboost.py [hunk 2]
```diff
@@ -434,26 +434,75 @@ def log_feature_importance_plot(features, importance, importance_type):
...
+                zip(feature_ylocs, importances_per_class_by_feature)
+            ):
+                for class_idx, (offset, class_importance) in enumerate(
+                    zip(offsets_per_yloc, importances_per_class)
+                ):
+                    (bar,) = ax.barh(
+                        feature_yloc + offset,
+                        class_importance,
+                        align="center",
+                        height=(0.5 / num_classes),
```
Comment (harupy) [inline mlflow/xgboost.py:492]:
```suggestion
                        height=(0.5 / (num_classes - 1)),
```

Do we need to divide by `num_classes - 1` here to fill the gap between bars?

feature_importance_weight

File: mlflow/xgboost.py [hunk 3]
```diff
@@ -434,26 +434,75 @@ def log_feature_importance_plot(features, importance, importance_type):
...
+            # Define offsets on the y-axis that are used to evenly space the bars for each class
+            # around the y-axis position of each feature
+            offsets_per_yloc = np.linspace(-0.5, 0.5, num_classes) / 2 if num_classes > 1 else [0]
+            for feature_idx, (feature_yloc, importances_per_class) in enumerate(
+                zip(feature_ylocs, importances_per_class_by_feature)
+            ):
+                for class_idx, (offset, class_importance) in enumerate(
+                    zip(offsets_per_yloc, importances_per_class)
+                ):
+                    (bar,) = ax.barh(
```
Comment (harupy) [inline mlflow/xgboost.py:488]: Nice unpacking :)

File: mlflow/xgboost.py [hunk 4]
```diff
@@ -434,26 +434,75 @@ def log_feature_importance_plot(features, importance, importance_type):
...
+                features = features[indices]
+                importances_per_class_by_feature = np.array(
+                    [[importance] for importance in importances_per_class_by_feature]
+                )
+                # In this case, do not include class labels on the feature importance plot because
+                # only one importance value has been provided per feature, rather than an
+                # one importance value for each class per feature
+                label_classes_on_plot = False
+            else:
+                label_classes_on_plot = True
```
Comment (harupy) [inline mlflow/xgboost.py:458]: Can we sort a 2D importance matrix (that linear boosters generates) as well?

```python
import numpy as np

features = np.array(["a", "b", "c"])
importance = [
    # class0, class1, class2
    [7, 8, 9],  # a
    [4, 5, 6],  # b
    [1, 2, 3],  # c
]
importances_per_class_by_feature = np.array(importance)
abs_sum = np.abs(importances_per_class_by_feature).sum(axis=1)
# or abs_mean = np.abs(importances_per_class_by_feature).mean(axis=1)
indices = np.argsort(abs_sum)

print(importances_per_class_by_feature[indices])
# [[1 2 3]
#  [4 5 6]
#  [7 8 9]]

print(features[indices])
# ['c' 'b' 'a']
```

File: mlflow/xgboost.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/xgboost.py]: Great suggestion! Done!

File: mlflow/xgboost.py (hunk 1 above)
Comment (dbczumar) [reply mlflow/xgboost.py]: Thanks for catching this! Done! Here's a screenshot from tree booster on MNIST:

feature_importance_weight

File: mlflow/xgboost.py (hunk 4 above)
Comment (dbczumar) [reply mlflow/xgboost.py]: Absolutely! Done! (Chose `sum()` for the magnitude metric rather than `mean()`). Here's a screenshot from a linear booster on MNIST:

feature_importance_weight

**Review by dbczumar — COMMENTED** [review]
@harupy Thanks for the awesome review feedback! I've addressed your comments. Can you take another look?

**Review by harupy — APPROVED** [review]
LGTM!
